### PR TITLE
Add first working version of the page

### PR DESCRIPTION
**Pull request by philss**
This ~commit~ PR introduces the first version of Broadway dashboard as an
additional page for the Phoenix LiveDashboard.

It works by recording events from the pipelines and displaying a graph
with each stage as a circle.
The circles turn red when they are more busy and green and they are less
busy.

This work is based on Marlus Saraiva's code from his presentation to
ElixirConf 2019. In that talk he presented a graph showing the
work of a Broadway pipeline, which is essentially the same from this
change. See: https://www.youtube.com/watch?v=tPu-P97-cbE

## Demo

https://user-images.githubusercontent.com/381213/119524701-40fa9a80-bd54-11eb-8dba-9da4d85c7a82.mp4

## TODO

- [x] high level changes based on your input
- [x] decide if we need to worry with complete restarts of the pipelines => yes
- [x] add GitHub actions setup
- [x] add tests
- [x] add docs
- [x] decide about license - should we go with Apache 2 like Broadway? or MIT like Phoenix?
- [x] wait **Phoenix Live Dashboard** for **Layered Graph Component**

### CODE REVIEW

File: config/config.exs [hunk 1]
```diff
@@ -0,0 +1,10 @@
+use Mix.Config
+
+config :phoenix, :json_library, Jason
+
+config :logger, level: :warn
+config :logger, :console, format: "[$level] $message\n"
+
+config :phoenix, :stacktrace_depth, 20
```
Comment (josevalim) [inline config/config.exs:8]:
```suggestion
config :phoenix, :json_library, Jason
config :phoenix, :stacktrace_depth, 20

config :logger, level: :warn
config :logger, :console, format: "[$level] $message\n"
```

File: dev.exs [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+defmodule DemoWeb.Endpoint do
+  use Phoenix.Endpoint, otp_app: :phoenix_live_dashboard
+
+  socket "/live", Phoenix.LiveView.Socket
+  socket "/phoenix/live_reload/socket", Phoenix.LiveReloader.Socket
+
+  plug Phoenix.LiveReloader
+  plug Phoenix.CodeReloader
+
+  plug Phoenix.LiveDashboard.RequestLogger,
```
Comment (josevalim) [inline dev.exs:149]: We probably don't need this one for this dev.

File: dev.exs [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+  plug Plug.Session,
+    store: :cookie,
+    key: "_live_view_key",
+    signing_salt: "/VEDsdfsffMnp5"
+
+  plug Plug.RequestId
+  plug DemoWeb.Router
+end
+
+Application.ensure_all_started(:os_mon)
```
Comment (josevalim) [inline dev.exs:162]: We probably don't need this one either. :)

File: lib/broadway_dashboard/pipeline_graph.ex [hunk 4]
```diff
@@ -0,0 +1,120 @@
+defmodule BroadwayDashboard.PipelineGraph do
```
Comment (josevalim) [inline lib/broadway_dashboard/pipeline_graph.ex:1]: Make sure we build this on the dashboard and only once!

File: lib/broadway_dashboard/metrics.ex [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+  defp ensure_server_started_at_node(target_node) do
+    case :rpc.call(target_node, Process, :whereis, [__MODULE__]) do
+      pid when is_pid(pid) ->
+        :ok
+
+      nil ->
+        with :ok <- Teleporter.teleport_metrics_code(target_node),
+             pid when is_pid(pid) <-
+               Node.spawn(target_node, __MODULE__, :start, [[from_node: node()]]) do
```
Comment (josevalim) [inline lib/broadway_dashboard/metrics.ex:36]: This can probably be :rpc.call.

File: lib/broadway_dashboard/metrics.ex [hunk 6]
```diff
@@ -0,0 +1,159 @@
+defmodule BroadwayDashboard.Metrics do
```
Comment (josevalim) [inline lib/broadway_dashboard/metrics.ex:1]: Have one metric process per pipeline.

File: lib/broadway_dashboard/telemetry.ex [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+      [:broadway, :processor, :message, :exception],
+      [:broadway, :consumer, :start],
+      [:broadway, :consumer, :stop],
+      [:broadway, :batcher, :start],
+      [:broadway, :batcher, :stop]
+    ]
+
+    id = {__MODULE__, parent}
+
+    :telemetry.attach_many(id, events, &handle_event/4, %{})
```
Comment (josevalim) [inline lib/broadway_dashboard/telemetry.ex:29]: Could we pass the counters as the fourth argument here?

File: lib/broadway_dashboard/counters.ex [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+                put_atomics(acc, stage_detail.name, stage_detail.concurrency)
+              end)
+          end
+        end)
+
+      # The counters are used for overall counting.
+      # Successful and error (2).
+      :persistent_term.put(
+        {__MODULE__, pipeline},
+        {:counters.new(2, [:write_concurrency]), processes_atomics}
```
Comment (josevalim) [inline lib/broadway_dashboard/counters.ex:63]: Let's do a single atomic!

**Comment by philss** [comment]
As we discussed, the improvements are going to be tackled in other PRs.

**Pull request merged**